### DISCUSSION TITLE
Query cache - trimming empty keys on object

### DISCUSSION THREAD

**Discussion by CutiePi** [TanStack/query#5307] [General]
Is it normal/expected? 

I have one use case where I need to preserve those keys

So for now I `JSON.stringify` the content before passing it to the cache. Otherwise, I lose all the empty keys. 
Couldn't find an option to preserve keys. 
Example in the object if there is `{key1: 1, anotherKey:{}` when back from cache it loses `anotherKey`

**Comment by TkDodo** [comment]
can you show an example? we use `JSON.stringify` under the hood to hash the key. So

```
{ key1: 1, anotherKey:{} }
```

should yield

```
'{"key1":1,"anotherKey":{}}'
```

```
{ key1: 1, anotherKey: undefined }
```

you'll get:

```
'{"key1":1}'
```

but that is what `JSON.stringify` does ...

**Reply by CutiePi** [reply]
Hey, there sure let me give you my example without using Stringify:
**We are on v3.13.0**

The get itself is just fetching data from a `.json` file

However, if I stringify the result before the cache (basically on the first call), it preserves all keys and values. 
Other wise this is what happens
 
first call
```JavaScript
{
    "state": {
        "models": {
            "key1": {},
            "key2": {},
            "key3": {},
            "key4": {},
            "key5": {},
            "key6": {},
            "key7": {},
            "key8": {},
            "key8": {},
            "key9": {},
            "key10": {},
            "key11": {},
            "key12": {},
            "key13": { actual stuff in here multi-level deep  }
            },
        }
    },
}
```

However, when we do subsequent uses of the hook we receive because it loads from cache
If I hard refresh the page then it loads the above 
second call
```JavaScript
{
    "state": {
        "models": {
            "key13": { actual stuff in here multi-level deep  }
            },
        }
    },
}
```

**Reply by TkDodo** [reply]
Okay sorry, I was thinking you were talking about Query Keys, but you are actually talking about data returned by useQuery? We shouldn't touch that. Whatever you return from the QueryFunction will be stored. Can you show this behavior in a codesandbox reproduction?

**Reply by CutiePi** [reply]
I'll try that over the weekend or Monday, will make a sandbox env to reproduce if I can.

**Reply by CutiePi** [reply]
I can't reproduce it in the sandbox. Here's a dev tools video capture on the app. You can see the value in the cache losing its empty keys. But it does not occur if I save the data as a JSON string. It seems to be triggered by a navigation page change. 

https://user-images.githubusercontent.com/17714927/234012437-82704ae6-4e15-405a-8227-f9e778c06159.mp4

And if I save the same data as a JSON string instead it stays untouched


https://user-images.githubusercontent.com/17714927/234018875-51a388f7-9ab8-4357-bf50-915b4c59301b.mp4
